### PR TITLE
[SPARK-39615][SQL] Make listColumns be compatible with 3 layer namespace

### PR DESCRIPTION
**Pull request by zhengruifeng**
### What changes were proposed in this pull request?
Make listColumns be compatible with 3 layer namespace

### Why are the changes needed?
for 3 layer namespace compatiblity

### Does this PR introduce _any_ user-facing change?
Yes

### How was this patch tested?
added UT

### CODE REVIEW

**Comment by zhengruifeng** [comment]
cc @cloud-fan @amaliujia

File: sql/core/src/main/scala/org/apache/spark/sql/internal/CatalogImpl.scala [hunk 1]
```diff
@@ -208,8 +209,18 @@ class CatalogImpl(sparkSession: SparkSession) extends Catalog {
...
   override def listColumns(tableName: String): Dataset[Column] = {
-    val tableIdent = sparkSession.sessionState.sqlParser.parseTableIdentifier(tableName)
-    listColumns(tableIdent)
+    // calling `sqlParser.parseTableIdentifier` to parse tableName. If it contains only table name
+    // and optionally contains a database name(thus a TableIdentifier), then we look up the table in
+    // sessionCatalog. Otherwise we try `sqlParser.parseMultipartIdentifier` to have a sequence of
+    // string as the qualified identifier and resolve the table through SQL analyzer.
+    try {
+      val tableIdent = sparkSession.sessionState.sqlParser.parseTableIdentifier(tableName)
+      listColumns(tableIdent)
```
Comment (amaliujia) [inline sql/core/src/main/scala/org/apache/spark/sql/internal/CatalogImpl.scala:218]: I think we only need one listColumns? 

If tableIdent is in SessionCatalog, then we convert it `spark_catalog.${tableIdent.database}.${tableIdent.name}`?

Comment (amaliujia) [reply sql/core/src/main/scala/org/apache/spark/sql/internal/CatalogImpl.scala:218]: Similarly as `getTable`: https://github.com/apache/spark/blob/b88cabb754073e69ce82fd561e6eac15046e633e/sql/core/src/main/scala/org/apache/spark/sql/internal/CatalogImpl.scala#L260

File: sql/core/src/main/scala/org/apache/spark/sql/internal/CatalogImpl.scala [hunk 2]
```diff
@@ -23,14 +23,15 @@ import scala.util.control.NonFatal
 import org.apache.spark.sql._
 import org.apache.spark.sql.catalog.{Catalog, CatalogMetadata, Column, Database, Function, Table}
 import org.apache.spark.sql.catalyst.{DefinedByConstructorParams, FunctionIdentifier, TableIdentifier}
-import org.apache.spark.sql.catalyst.analysis.{ResolvedNamespace, ResolvedTable, ResolvedView, UnresolvedDBObjectName, UnresolvedNamespace, UnresolvedTable, UnresolvedTableOrView}
+import org.apache.spark.sql.catalyst.analysis._
 import org.apache.spark.sql.catalyst.catalog._
 import org.apache.spark.sql.catalyst.encoders.ExpressionEncoder
-import org.apache.spark.sql.catalyst.plans.logical.{CreateTable, LocalRelation, RecoverPartitions, ShowTables, SubqueryAlias, TableSpec, View}
+import org.apache.spark.sql.catalyst.plans.logical._
```
Comment (amaliujia) [inline sql/core/src/main/scala/org/apache/spark/sql/internal/CatalogImpl.scala:29]: revert

**Review by cloud-fan — APPROVED** [review]

**Review by amaliujia — APPROVED** [review]
LGTM!

**Pull request closed without merging**